cmd/deploy/input_test.go

```diff
@@ -2,9 +2,240 @@ package main
+	type fixture struct {
+		name       string
+		rawLookups map[string]string
+		want       want
+		wantErr    string
+	}
+
+	for _, x := range []fixture{
+		{
+			"happy path",
+			map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "example:foo",
+				"sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86": "foo/bar:baz",
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": "localhost:5000/example:bar",
+			},
+			want{
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				map[ociregistry.Digest]registry.Reference{
+					"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": {
+						Host:       "docker.io",
+						Repository: "library/example",
+						Tag:        "foo",
+						Digest:     "sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+					},
+					"sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86": {
+						Host:       "docker.io",
+						Repository: "foo/bar",
+						Tag:        "baz",
+						Digest:     "sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86",
+					},
+					"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": {
+						Host:       "localhost:5000",
+						Repository: "example",
+						Tag:        "bar",
+						Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+					},
+				},
+			},
+			"",
+		},
+		{
+			"bad digest",
+			map[string]string{
+				"invalid digest": "example:foo",
+			},
+			want{},
+			"lookup key \"invalid digest\" invalid: invalid checksum digest format",
+		},
+		{
+			"bad reference",
+			map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "://#",
+			},
+			want{},
+			"failed to parse lookup ref \"://#\": invalid reference syntax (\"://#\")",
+		},
+	} {
+		t.Run(x.name, func(t *testing.T) {
+			lookupDigest, lookups, err := getLookup(x.rawLookups)
+
+			if x.wantErr != "" && (err == nil || err.Error() == "") {
+				t.Fatalf("Expected error not returned: %s", x.wantErr)
+			}
+
+			if x.wantErr != "" && err.Error() != x.wantErr {
+				t.Fatalf("Expected error doesn't match. got:\n'%s',\n\nexpected:\n'%s'", err.Error(), x.wantErr)
+			}
+
+			if lookupDigest != x.want.digest {
+				t.Errorf("Expected last digest. got:\n%s\n\nexpected:\n%s\n", lookupDigest, x.want.digest)
+			}
+
+			if len(lookups) != len(x.want.lookups) {
+				t.Errorf("Reference count mismatch. got: %d, expected: %d", len(lookups), len(x.want.lookups))
+			}
+
+			for digest, ref := range lookups {
+				if _, ok := x.want.lookups[digest]; !ok {
+					t.Errorf("digest missing from lookups: %s", digest)
+				} else if ref != x.want.lookups[digest] {
+					gotJson, _ := json.MarshalIndent(MarshallableReference(ref), "", "  ")
+					expectedJson, _ := json.MarshalIndent(MarshallableReference(x.want.lookups[digest]), "", "  ")
```

Comment:
Even reading this, I'm still confused :sweat_smile: couldn't this just be `%+v` or `%#v` in the error message?